mqtt/handler_test.go

```diff
@@ -0,0 +1,406 @@
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestAuthPublish(t *testing.T) {
+	handler := newHandler()
+	var payload = []byte("[{'n':'test-name', 'v': 1.2}]")
+
+	chID, err := idProvider.ID()
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	var topic = "channels/" + chID + "/messages"
+
+	cases := []struct {
+		desc    string
+		client  *session.Client
+		err     error
+		topic   *string
+		payload []byte
+	}{
+		{
+			desc:    "publish with inactive client",
+			client:  nil,
+			err:     mqtt.ErrClientNotInitialized,
+			topic:   &topic,
+			payload: payload,
+		},
+		{
+			desc:    "publish without topic",
+			client:  &sessionClient,
+			err:     mqtt.ErrMissingTopicPub,
+			topic:   nil,
+			payload: payload,
+		},
+		{
+			desc:    "publish with malformed topic",
+			client:  &sessionClient,
+			err:     mqtt.ErrMalformedTopic,
+			topic:   &invalidTopic,
+			payload: payload,
+		},
+	}
+
+	for _, tc := range cases {
+		err := handler.AuthPublish(tc.client, tc.topic, &tc.payload)
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestAuthSubscribe(t *testing.T) {
+	handler := newHandler()
+	var invalidTopics = []string{invalidTopic}
+	var topics = []string{"channels/" + chanID + "/messages"}
+	invalidChanIDTopics := []string{"channels/" + invalidChanID + "/messages"}
+
+	cases := []struct {
+		desc   string
+		client *session.Client
+		err    error
+		topic  *[]string
+	}{
+		{
+			desc:   "subscribe without active session",
+			client: nil,
+			err:    mqtt.ErrClientNotInitialized,
+			topic:  &topics,
+		},
+		{
+			desc:   "subscribe without topics",
+			client: &sessionClient,
+			err:    mqtt.ErrMissingTopicSub,
+			topic:  nil,
+		},
+		{
+			desc:   "subscribe with invalid topics",
+			client: &sessionClient,
+			err:    mqtt.ErrMalformedTopic,
+			topic:  &invalidTopics,
+		},
+		{
+			desc:   "subscribe with invalid channel ID",
+			client: &sessionClient,
+			err:    mqtt.ErrAuthentication,
+			topic:  &invalidChanIDTopics,
+		},
+		{
+			desc:   "subscribe with invalid thing ID",
+			client: &invalidThingSessionClient,
+			err:    mqtt.ErrAuthentication,
+			topic:  &topics,
+		},
+		{
+			desc:   "subscribe with active session and valid topics",
+			client: &sessionClient,
+			err:    nil,
+			topic:  &topics,
+		},
+	}
+
+	for _, tc := range cases {
+		err := handler.AuthSubscribe(tc.client, tc.topic)
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
+	}
+}
+
+func TestConnect(t *testing.T) {
+	handler := newHandler()
+	buf.Reset()
+	cases := []struct {
+		desc     string
+		client   *session.Client
+		expected string
+	}{
+		{
+			desc:     "connect without active session",
+			client:   nil,
+			expected: fmt.Sprint(errors.Wrap(mqtt.ErrFailedConnect, mqtt.ErrClientNotInitialized).Error()),
```

Comment:
use `err: errors.Wrap(mqtt.ErrFailedConnect, mqtt.ErrClientNotInitialized)` here and extract the string latter